File: chapter_3_stacks_and_queues/python/3.2-stack_min.py
Stack = __import__('stack').Stack


class StackWithMin(Stack):
    def __init__(self):
        super().__init__()
        # new stack with minimums
        self.s2 = Stack()

    def push(self, value):
        if value < self.min():
            self.s2.push(value)
        super().push(value)

    def pop(self):
        # If min is getting popped, remove from second list also
        tmp = super().pop()
        if tmp == self.min():
            self.s2.pop()
        return tmp

    def min(self):
        if self.s2.is_empty():
            return float('inf')
        return self.s2.peek()


if __name__ == 'main':
    swm = StackWithMin()
    print(swm.__dict__)
    # {'stack': [], 'length': 0, 's2': <stack.Stack objek object at 0x7f6a0681a390>}
    for i in [4, -1, 2, 6, 2, -5]:
        swm.push(i)
    print(swm.min())  # -5
    print(swm.stack)  # [4, -1, 2, 6, 2, -5]
    print(swm.s2.stack)  # [4, -1, -5]
    print(swm.pop())  # -5
    print(swm.stack)  # [4, -1, 2, 6, 2]
    print(swm.s2.stack)  # [4, -1]

File: chapter_3_stacks_and_queues/js/3.2-stackMin.js
const { Stack } = require("./stack")

class StackWithMin extends Stack {
  constructor() {
    super()
    /* new stack with minimums */
    this.s2 = new Stack()
  }

  push(value) {
    if (value <= this.min()) this.s2.push(value)
    super.push(value)
  }

  pop() {
    const value = super.pop()
    /* If min is getting popped, remove from second list also */
    if (value == this.min()) this.s2.pop()
    return value
  }

  min() {
    return this.s2.isEmpty() ? Number.MAX_SAFE_INTEGER : this.s2.peek()
  }
}

swm = new StackWithMin()
console.log(swm)
/*
StackWithMin {
  stack: [],
  length: 0,
  s2: Stack { stack: [], length: 0 } }
*/

console.log(swm.min()) // 9007199254740991
swm.push(5)
swm.push(10)
swm.push(-15)
swm.push(20)
console.log(swm)
/* 
StackWithMin {
  stack: [ 5, 10, -15, 20 ],
  length: 4,
  s2: Stack { stack: [ 5, -15 ], length: 2 } }
*/
console.log(swm.min()) // -15
console.log(swm.pop()) // 20
console.log(swm.pop()) // -15
console.log(swm)
/* 
StackWithMin {
  stack: [ 5, 10 ],
  length: 2,
  s2: Stack { stack: [ 5 ], length: 1 } }
*/
